autofaiss/external/distributed.py

```diff
@@ -0,0 +1,367 @@
+"""
+Building the index with pyspark.
+"""
+import math
+import multiprocessing
+import os
+from functools import partial
+from multiprocessing.pool import ThreadPool
+from tempfile import NamedTemporaryFile, TemporaryDirectory
+from typing import List, Optional, Union, Iterator, Tuple, Callable, Any
+
+import faiss
+import fsspec
+import numpy as np
+import pandas as pd
+from fsspec import get_filesystem_class
+from fsspec.implementations.hdfs import PyArrowHDFS
+from tqdm import tqdm
+
+from autofaiss.readers.embeddings_iterators import get_file_list, get_matrix_reader
+from autofaiss.utils.cast import cast_memory_to_bytes
+from autofaiss.utils.decorators import Timeit
+
+
+def _yield_embeddings_batch(
+    embeddings_paths: List[str],
+    chunk_sizes: List[int],
+    file_system: fsspec.AbstractFileSystem,
+    start: int,
+    end: int,
+    embedding_column_name: str,
+    file_format: str,
+    id_columns: Optional[List[str]] = None,
+):
+    """Lazy load a batch of embeddings."""
+    if isinstance(file_system, PyArrowHDFS):
+        file_system = get_filesystem_class("hdfs")()
+    cur_start = cur_end = 0
+    for chunk_size, file_path in zip(chunk_sizes, embeddings_paths):
+        cur_end += chunk_size
+        if cur_end < start:
+            cur_start += chunk_size
+            continue
+        slice_start = max(0, start - cur_start)
+        slice_end = min(chunk_size, end - cur_start)
+        with get_matrix_reader(file_format, file_system, file_path, embedding_column_name, id_columns) as matrix_reader:
+            yield matrix_reader.get_lazy_array().get_rows(start=slice_start, end=slice_end)
+        if cur_end > end:
+            break
+        cur_start += chunk_size
+
+
+def _get_index_from_bytes(index_bytes: Union[bytearray, bytes]) -> faiss.Index:
+    """Transforms a bytearray containing a faiss index into the corresponding object."""
+
```

Comment:
Maybe you could put it here? https://github.com/criteo/autofaiss/blob/master/autofaiss/indices/index_utils.py